### PR TITLE
Speechlm 3b shehzeen

### PR DESCRIPTION
**Pull request by shehzeen**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py [hunk 1]
```diff
@@ -124,14 +156,6 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
                 self.frozen_model.enc_dec_model.decoder_cfg.hidden_size, speech_codebook_size
             )
 
-        self.speech_offset = speech_offset
```
Comment (subhankar-ghosh) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py:127]: Why was all of these initializations removed from here?

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_pretrain_model.py [hunk 2]
```diff
@@ -105,6 +106,21 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
         speech_offset = cfg.data.get('speech_offset', 30000)
         speech_head_type = cfg.get('speech_head_type', 'token_level')  # token_level, linear
 
+        self.speech_offset = speech_offset
```
Comment (subhankar-ghosh) [inline nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_pretrain_model.py:109]: why is there two different T5 models? One for pretrain and one for SFT I am assuming. Why does the model need to be defined separately?

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py (hunk 1 above)
Comment (shehzeen) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_model.py:127]: The initializations have been moved above (Line 104 to Line 110). I believe some of the variables are being used in initializing speech heads thats why they were moved above.

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_pretrain_model.py (hunk 2 above)
Comment (shehzeen) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_pretrain_model.py:109]: Yes, as discussed in Slack, these two files can potentially be merged in the future once the two models are stable.

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py [hunk 3]
```diff
@@ -673,10 +669,16 @@ def forward(
...
                                 )
                             elif self.speech_head_type == "linear":
-                                last_layer_logits = self.speech_tokens_heads[i](dec_output)
+                                last_layer_logits = self.speech_tokens_heads[i](dec_output)[0]
                             else:
                                 raise ValueError(f"Speech head type {self.speech_head_type} not supported")
-                            speech_logits[:, :, :, i] = last_layer_logits
+                            speech_logits_list.append(last_layer_logits)
+                            if self.parallel_output:
+                                speech_logits[:, :, :, i] = tensor_parallel.gather_from_tensor_model_parallel_region(
```
Comment (blisc) [inline nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:677]: Do you need this gather? If it's only for eval or inference, put it instead a if statement that only runs during eval

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_pretrain_model.py (hunk 2 above)
Comment (blisc) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_pretrain_model.py:109]: @paarthneekhara You should consider inheriting as many common functions as you can, and only overwrite those that you are changing

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py (hunk 3 above)
Comment (paarthneekhara) [reply nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py:677]: It was also being used during in the training step for logging audio on tensorboard. I believe we can move all that to only the validation step and add if statement here to work only in the validation mode. Should I make this change and push to this branch @blisc ?

File: nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_pretrain_model.py (hunk 2 above)
Comment (paarthneekhara) [reply nemo/collections/nlp/models/language_modeling/megatron_t5_speechlm_pretrain_model.py:109]: Yes, I think right now these two files became substantially different. The pretraining model is more like megatron_gpt_model.py and SFT model is more like megatron_gpt_prompt_learning.py . But yes, once we finalize the two models we can discuss how to merge them into one.

**Comment by subhankar-ghosh** [comment]
LGTM for now, we need to merge the two model classes in the near future once the models have stabilized.

**Review by subhankar-ghosh — APPROVED** [review]

File: nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py (hunk 3 above)
Comment (paarthneekhara) [reply nemo/collections/nlp/modules/common/megatron/token_level_encoder_decoder.py]: Resolved.

**Pull request merged**